src/main/scala/edu/uci/eecs/spectralLDA/datamoments/DataCumulantSketch.scala

```diff
@@ -15,83 +18,119 @@ import scala.collection.mutable
-                      sketcher: TensorSketcher[Double, Double])
+                      documents: RDD[(Long, SparseVector[Double])],
+                      sketcher: TensorSketcher[Double, Double],
+                      randomisedSVD: Boolean = true)
+                     (implicit tolerance: Double = 1e-9)
   : DataCumulantSketch = {
     val sc: SparkContext = documents.sparkContext
-    val dimVocab = documents.take(1)(0)._3.length
-    val numDocs = documents.count()
 
-    println("Start calculating first order moments...")
-    val M1: DenseVector[Double] = documents map {
-      case (_, length, vec) => update_firstOrderMoments(dimVocab, vec.toDenseVector, length)
-    } reduce (_ + _)
-
-    val firstOrderMoments: DenseVector[Double] = M1 / numDocs.toDouble
-    println("Finished calculating first order moments.")
+    val validDocuments = documents
+      .map {
+        case (id, wc) => (id, sum(wc), wc)
+      }
+      .filter {
+        case (_, len, _) => len >= 3
+      }
 
-    val (thirdOrderMoments: DenseMatrix[Double], unwhiteningMatrix: DenseMatrix[Double]) = computeThirdOrderMoments(
-      sc, alpha0, dimVocab, dimK,
-      numDocs, firstOrderMoments, documents,
-      tolerance
-    )
+    val dimVocab = validDocuments.take(1)(0)._3.length
+    val numDocs = validDocuments.count()
 
-    val thirdOrderMomentsSketch = sketcher.sketch(thirdOrderMoments)
-    new DataCumulantSketch(thirdOrderMomentsSketch, unwhiteningMatrix)
-  }
+    println("Start calculating first order moments...")
+    val firstOrderMoments: DenseVector[Double] = validDocuments
+      .map {
+        case (_, length, vec) => vec.toDenseVector / length
+      }
+      .reduce(_ + _)
+      .map(_ / numDocs.toDouble)
+    println("Finished calculating first order moments.")
 
-  private def computeThirdOrderMoments(sc: SparkContext,
-                                       alpha0: Double,
-                                       dimVocab: Int, dimK: Int,
-                                       numDocs: Long,
-                                       firstOrderMoments: DenseVector[Double],
-                                       documents: RDD[(Long, Double, SparseVector[Double])],
-                                       tolerance: Double)
-  : (DenseMatrix[Double], DenseMatrix[Double]) = {
     println("Start calculating second order moments...")
-    val (eigenVectors: DenseMatrix[Double], eigenValues: DenseVector[Double]) = whiten(sc, alpha0,
-      dimVocab, dimK, numDocs, firstOrderMoments, documents)
+    val (eigenVectors: DenseMatrix[Double], eigenValues: DenseVector[Double]) = if (randomisedSVD) {
+      whiten(sc, alpha0,
+        dimVocab, dimK, numDocs, firstOrderMoments, validDocuments)
+    }
+    else {
+      val E_x1_x2: DenseMatrix[Double] = validDocuments
+        .map { case (_, len, vec) => (len, vec.toDenseVector) }
+        .map { case (len, vec) => (vec * vec.t - diag(vec)) / (len * (len - 1)) }
```

Comment:
I believe those can be merged as well. An intermediate inner val is less costly than the 2 maps (until Scala does map fusion).